### PR TITLE
5080 r32 get contract periods

### PR DESCRIPTION
**Pull request by AnumehaSrivastava05**
Includes R40 changes

### CODE REVIEW

File: backend/src/main/java/ca/bc/gov/hlth/hnweb/controller/MspContractsController.java [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+		addAffectedParty(transaction, IdentifierType.PHN, getContractPeriodsRequest.getPhn());
+		return transaction;
+	}
+
+	private void auditGetContractPeriodsEnd(Transaction transaction, GetContractPeriodsResponse getContractPeriodsResponse) {
+		List<String> auditedPhns = new ArrayList<>();
+		List<String> auditedGroupNumbers = new ArrayList<>();
+
+		transactionComplete(transaction);		
+		addAffectedParty(transaction, IdentifierType.PHN, getContractPeriodsResponse.getPhn());
```
[the remark this answers is not in the document]
Comment (daveb-hni) [reply backend/src/main/java/ca/bc/gov/hlth/hnweb/controller/MspContractsController.java:116]: Is being confirmed, will update later if any changes needed.

File: backend/src/main/java/ca/bc/gov/hlth/hnweb/controller/MspContractsController.java [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+		addAffectedParty(transaction, IdentifierType.PHN, contractInquiryRequest.getPhn());
+		addAffectedParty(transaction, IdentifierType.GROUP_NUMBER, contractInquiryRequest.getGroupNumber());
+		return transaction;
+	}
+
+	private void auditContractInquiryEnd(Transaction transaction, ContractInquiryResponse contractInquiryResponse) {
+		List<String> auditedPhns = new ArrayList<>();
+
+		transactionComplete(transaction);		
+		addAffectedParty(transaction, IdentifierType.PHN, contractInquiryResponse.getPhn());
```
[the remark this answers is not in the document]
Comment (daveb-hni) [reply backend/src/main/java/ca/bc/gov/hlth/hnweb/controller/MspContractsController.java:146]: Is being confirmed, will update later if any changes needed.

File: backend/src/main/java/ca/bc/gov/hlth/hnweb/converter/rapid/RPBSPMC0Converter.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+				beneficiaryContractPeriod.setPhn(b.getPhn());
+				beneficiaryContractPeriod.setFirstName(StringUtils.trim(b.getFirstName()));
+				beneficiaryContractPeriod.setLastName(StringUtils.trim(b.getLastName()));
+				beneficiaryContractPeriod.setDateOfBirth(convertDate(b.getBirthDate()));
+				beneficiaryContractPeriod.setGender(b.getGender());
+				beneficiaryContractPeriod.setContractHolder(cp.getPhn());
+				beneficiaryContractPeriod.setGroupNumber(cp.getGroupNumber());
+				beneficiaryContractPeriod.setRelationship(cp.getRelationship());
+				beneficiaryContractPeriod.setEffectiveDate(convertDate(cp.getEffectiveDate()));
+				beneficiaryContractPeriod.setCancelDate(StringUtils.equals("0000-00-00", cp.getCancelDate()) ? null : convertDate(cp.getCancelDate()));
```
[the remark this answers is not in the document]
Comment (daveb-hni) [reply backend/src/main/java/ca/bc/gov/hlth/hnweb/converter/rapid/RPBSPMC0Converter.java]: Yup, done, had already but hadn't pushed the change

File: backend/src/main/java/ca/bc/gov/hlth/hnweb/security/SecurityConfig.java [hunk 4]
```diff
@@ -58,9 +58,9 @@ protected void configure(HttpSecurity http) throws Exception {
             .mvcMatchers(HttpMethod.POST,"/eligibility/check-msp-coverage-status").hasRole("E45")
             .mvcMatchers(HttpMethod.POST,"/eligibility/check-eligibility").hasRole("R15")
             .mvcMatchers(HttpMethod.POST,"/eligibility/inquire-phn").hasRole("R41")
-            .mvcMatchers(HttpMethod.POST, "/group-member/**").fullyAuthenticated()
             .mvcMatchers(HttpMethod.POST,"/enrollment/**").fullyAuthenticated()
             .mvcMatchers(HttpMethod.POST,"/group-member/**").fullyAuthenticated()
+            .mvcMatchers(HttpMethod.POST,"/msp-contracts/**").fullyAuthenticated()
```
[the remark this answers is not in the document]
Comment (daveb-hni) [reply backend/src/main/java/ca/bc/gov/hlth/hnweb/security/SecurityConfig.java:63]: Added space on new entry, I'll check on merge that it is consistent and resolve any conflicts then also.

File: backend/src/main/java/ca/bc/gov/hlth/hnweb/security/TransactionType.java [hunk 5]
```diff
@@ -16,6 +16,8 @@
 	UPDATE_NUMBER_AND_OR_DEPT("UpdateNumberAnd/OrDept"),
 	CANCEL_GROUP_MEMBER("CancelGroupMember"),
 	CANCEL_DEPENDENT("CancelDependent"),
+	GET_CONTRACT_PERIODS("Get Contract Periods"),
```
[the remark this answers is not in the document]
Comment (daveb-hni) [reply backend/src/main/java/ca/bc/gov/hlth/hnweb/security/TransactionType.java:19]: Fixed spacing

File: backend/src/main/java/ca/bc/gov/hlth/hnweb/service/MspContractsService.java [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+/**
+ * Service for:
+ *  Get Contract Periods. (R32), ContractInquiry(R40)
+ */
+@Service
+public class MspContractsService extends BaseService {
+
+	private static final Logger logger = LoggerFactory.getLogger(MspContractsService.class);
+
+	public static final String TRANSACTION_ID = "TransactionID";
```
[the remark this answers is not in the document]
Comment (daveb-hni) [reply backend/src/main/java/ca/bc/gov/hlth/hnweb/service/MspContractsService.java:30]: Moved to BaseService  and refactored all services

File: backend/src/test/java/ca/bc/gov/hlth/hnweb/controller/MspContractsControllerTest.java [hunk 7]
```diff
@@ -0,0 +1,391 @@
...
+     * @param registry
+     */
+    @DynamicPropertySource
+    static void registerMockUrlProperty(DynamicPropertyRegistry registry) {
+        registry.add("rapid.url", () -> String.format("http://localhost:%s", mockBackEnd.getPort()));
+    }
+	
+}
+
+
```
[the remark this answers is not in the document]
Comment (daveb-hni) [reply backend/src/test/java/ca/bc/gov/hlth/hnweb/controller/MspContractsControllerTest.java:391]: Removed

File: backend/src/test/java/ca/bc/gov/hlth/hnweb/controller/MspContractsControllerTest.java [hunk 8]
```diff
@@ -0,0 +1,391 @@
...
+        List<BeneficiaryContractPeriod> beneficiarieBeneficiaryContractPeriods = getContractPeriodsResponse.getBeneficiaryContractPeriods();
+        assertEquals(4, beneficiarieBeneficiaryContractPeriods.size());
+       
+        //The results consist of two groups of MSP Contract Periods, Beneficiaries PHN 9873672248 and PHN 9873672255.        
+        
+		/**
+		 * 9873672248SPBIGDATASNAME                     SPBIGDATAFNAME                               1983-01-01F
+		 * 98736722550000001S2022-02-010000-00-00 
+		 * 98736722484044574C2022-02-010000-00-00                                                                                                                                                                                                                                                                                                                         
+		 */
```
[the remark this answers is not in the document]
Comment (daveb-hni) [reply backend/src/test/java/ca/bc/gov/hlth/hnweb/controller/MspContractsControllerTest.java:79]: Changed to use non public comments, that was making it look odd, but will keep the multi line as single line made it less readable when I tried them.

File: frontend/src/components/mspcontracts/BeneficiaryContractPeriod.vue [hunk 9]
```diff
@@ -0,0 +1,29 @@
+<template>
+  <td>{{ beneficiaryContractPeriod.phn }}</td>
+  <td>{{ fullName }}</td>
+  <td>{{ beneficiaryContractPeriod.contractHolder }}</td>
+  <td>{{ beneficiaryContractPeriod.relationship }}</td>
+  <td>{{ beneficiaryContractPeriod.groupNumber }}</td>
+  <td>{{ beneficiaryContractPeriod.effectiveDate }}</td>
+  <td>{{ beneficiaryContractPeriod.cancelDate }}</td>
+  <td>{{ beneficiaryContractPeriod.cancelReason === 'E' ? 'Eligible' : '' }}</td>
```
[the remark this answers is not in the document]
Comment (daveb-hni) [reply frontend/src/components/mspcontracts/BeneficiaryContractPeriod.vue:9]: Extracted to computed

File: frontend/src/components/mspcontracts/BeneficiaryContractPeriod.vue [hunk 10]
```diff
@@ -0,0 +1,29 @@
+<template>
+  <td>{{ beneficiaryContractPeriod.phn }}</td>
+  <td>{{ fullName }}</td>
+  <td>{{ beneficiaryContractPeriod.contractHolder }}</td>
+  <td>{{ beneficiaryContractPeriod.relationship }}</td>
```
[the remark this answers is not in the document]
Comment (daveb-hni) [reply frontend/src/components/mspcontracts/BeneficiaryContractPeriod.vue]: Extracted to common mapping

File: backend/src/main/java/ca/bc/gov/hlth/hnweb/controller/MspContractsController.java [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+	private void auditGetContractPeriodsEnd(Transaction transaction, GetContractPeriodsResponse getContractPeriodsResponse) {
+		List<String> auditedPhns = new ArrayList<>();
+		List<String> auditedGroupNumbers = new ArrayList<>();
+
+		transactionComplete(transaction);		
+		addAffectedParty(transaction, IdentifierType.PHN, getContractPeriodsResponse.getPhn());
+		auditedPhns.add(getContractPeriodsResponse.getPhn());
+		
+		getContractPeriodsResponse.getBeneficiaryContractPeriods().forEach(bcp -> {
+			if(!auditedPhns.contains(bcp.getPhn())) {
```
[the remark this answers is not in the document]
Comment (daveb-hni) [reply backend/src/main/java/ca/bc/gov/hlth/hnweb/controller/MspContractsController.java:120]: Added spacing

File: backend/src/main/java/ca/bc/gov/hlth/hnweb/controller/MspContractsController.java [hunk 12]
```diff
@@ -0,0 +1,156 @@
...
+
+	private void auditContractInquiryEnd(Transaction transaction, ContractInquiryResponse contractInquiryResponse) {
+		List<String> auditedPhns = new ArrayList<>();
+
+		transactionComplete(transaction);		
+		addAffectedParty(transaction, IdentifierType.PHN, contractInquiryResponse.getPhn());
+		auditedPhns.add(contractInquiryResponse.getPhn());
+		
+		contractInquiryResponse.getContractInquiryBeneficiaries().forEach(cib -> {
+			if(!auditedPhns.contains(cib.getPhn())) {
```
[the remark this answers is not in the document]
Comment (daveb-hni) [reply backend/src/main/java/ca/bc/gov/hlth/hnweb/controller/MspContractsController.java:150]: Added spacing

File: frontend/src/components/mspcontracts/BeneficiaryContractPeriod.vue [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+  <td>{{ beneficiaryContractPeriod.groupNumber }}</td>
+  <td>{{ beneficiaryContractPeriod.effectiveDate }}</td>
+  <td>{{ beneficiaryContractPeriod.cancelDate }}</td>
+  <td>{{ beneficiaryContractPeriod.cancelReason === 'E' ? 'Eligible' : '' }}</td>
+</template>
+<script>
+export default {
+  props: {
+    beneficiaryContractPeriod: Object,
```
[the remark this answers is not in the document]
Comment (daveb-hni) [reply frontend/src/components/mspcontracts/BeneficiaryContractPeriod.vue:16]: Really prefer slightly verbose to ambiguous, could do an acronym, bcp, if that's better?

[the remark this answers is not in the document]
Comment (weskubo-cgi) [reply frontend/src/components/mspcontracts/BeneficiaryContractPeriod.vue:16]: I think bcp could be confusing for the caller. I think it's just confusing to have the Component and prop have the same name. contractPeriod is probably a good compromise.

[the remark this answers is not in the document]
Comment (daveb-hni) [reply frontend/src/components/mspcontracts/BeneficiaryContractPeriod.vue:16]: contractPeriod reads well, will go with that.

**Review by weskubo-cgi — APPROVED** [review]
Changes look good.

**Pull request merged**